Firefox build failures on OS X when clang preprocessor output is compiled

When building Firefox for OS X using sccache and clang, there is at least one compilation unit that fails complaining of duplicated symbols. I disabled the check for -Werror that allows compiling from preprocessor output. It rebuilt successfully.

File that failed: media/libav/libavcodec/avfft.c

details: macOS 10.12.4, latest sccache, mozilla-inbound @ c29534e2e891

Sccache log snippets:
[sccache_build_failure_sccache_log_snippets.txt](https://github.com/mozilla/sccache/files/972393/sccache_build_failure_sccache_log_snippets.txt)

Mozconfig:
[sccache.opt.mozconfig.txt](https://github.com/mozilla/sccache/files/972384/sccache.opt.mozconfig.txt)

Portion of the build log in question:
[sccache_build_failure.log.txt](https://github.com/mozilla/sccache/files/972387/sccache_build_failure.log.txt)

The results of ./mach build-backend -b CompileDB
[with_sccache_compile_commands.json.txt](https://github.com/mozilla/sccache/files/972378/with_sccache_compile_commands.json.txt)

I am using the S3 storage backend against a network local instance of Minio.

I just hit this locally, I hadn't tried to build Firefox with sccache on OS X in a while. The problem is that this file is compiled with `-include avfft_perms.h`, which does a `#include "avfft.h"`. sccache preprocesses the source file, then tries to compile it from the preprocessed source, but passes `-include avfft_perms.h` again, which results in the duplicate definitions. I suspect this works in CI because we build with `--enable-warnings-as-errors` there, and we have special handling for falling back to compiling from the original source when we fail to compile and `-Werror` is in the commandline.

Here's a simpler testcase that reproduces the issue:
```
luser@eye7:/tmp/sccache-test$ touch test.c
luser@eye7:/tmp/sccache-test$ echo "struct S {};" > test.h
luser@eye7:/tmp/sccache-test$ /build/sccache/target/x86_64-unknown-linux-musl/release/sccache /build/clang+llvm-4.0.0-x86_64-linux-gnu-ubuntu-16.10/bin/clang -o test.o -c test.c -include test.h
In file included from test.c:1:
In file included from <built-in>:1:
./test.h:1:8: error: redefinition of 'S'
struct S {};
       ^
./test.h:1:8: note: previous definition is here
struct S {};
       ^
1 error generated.
```

--- a/src/compiler/clang.rs
+++ b/src/compiler/clang.rs
@@ -173,8 +173,8 @@ mod test {
         assert_map_contains!(a.outputs, ("obj", PathBuf::from("foo.o")));
         //TODO: fix assert_map_contains to assert no extra keys!
         assert_eq!(1, a.outputs.len());
-        assert!(a.preprocessor_args.is_empty());
-        assert_eq!(ovec!["-arch", "xyz", "-fabc", "-I", "include", "-include", "file"], a.common_args);
+        assert_eq!(ovec!["-include", "file"], a.preprocessor_args);
+        assert_eq!(ovec!["-arch", "xyz", "-fabc", "-I", "include"], a.common_args);
     }
 
     #[test]

--- a/src/compiler/gcc.rs
+++ b/src/compiler/gcc.rs
@@ -144,9 +144,14 @@ fn _parse_arguments(arguments: &[OsString],
                 // -MF and -MQ are in this set but are handled separately
                 // because they are also preprocessor options.
                 a if argument_takes_value(a) => {
-                    common_args.push(arg.clone());
+                    let args = if a == "-include" {
+                        &mut preprocessor_args
+                    } else {
+                        &mut common_args
+                    };
+                    args.push(arg.clone());
                     if let Some(arg_val) = it.next() {
-                        common_args.push(arg_val);
+                        args.push(arg_val);
                     }
                 },
                 "-MF" |
@@ -557,8 +562,8 @@ mod test {
         assert_map_contains!(outputs, ("obj", PathBuf::from("foo.o")));
         //TODO: fix assert_map_contains to assert no extra keys!
         assert_eq!(1, outputs.len());
-        assert!(preprocessor_args.is_empty());
-        assert_eq!(ovec!["-fabc", "-I", "include", "-include", "file"], common_args);
+        assert_eq!(ovec!["-include", "file"], preprocessor_args);
+        assert_eq!(ovec!["-fabc", "-I", "include"], common_args);
         assert!(!msvc_show_includes);
     }
 

